### PR TITLE
A pure java alternative for power operations with integer positive exponents

### PR DESCRIPTION
**Pull request by schmittjoaopedro**
Changes proposed in this PR:
- Add a pure java alternative for power operations with integer positive exponents.
- Add tests to increase the coverage of expected results for power operations (some of these tests are waiting the release of the contraction ratio parameter in ibex)
 
@chocoteam/core-developer

### CODE REVIEW

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import org.chocosolver.solver.variables.events.RealEventType;
+import org.chocosolver.util.ESat;
+import org.chocosolver.util.objects.RealInterval;
+import org.chocosolver.util.tools.RealUtils;
+
+/**
+ * It's a pure java alternative to ibex that must be used to calculate
+ * power operations (y = x^k) with integer positive exponents (k).
+ *
+ * <p>
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:17]: The 3 following lines are out-dated, there is no reference to `jinterval` anymore, if I'm correct.

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ * This implementation is based from the chapter 16.1 "From Discrete to
+ * Continuous Constraints", example 16.2 [1].
+ * <p>
+ * [1] Francesca Rossi, Peter van Beek, and Toby Walsh. 2006. Handbook of
+ * Constraint Programming. Elsevier Science Inc., USA.
+ *
+ * @author João Pedro Schmitt
+ * @since 08/09/2020
+ */
+public class RealPowerPropagator extends Propagator<RealVar> {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:30]: In Choco, we name propagator with prefix `Prop`.
You could change the name to `PropRealPower`.

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    private RealVar yVar;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:32]: can be final

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    private RealVar xVar;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:34]: can be final

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    private int k;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:36]: can be final

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        return RealEventType.BOUND.getMask();
+    }
+
+    @Override
+    public ESat isEntailed() {
+        ESat result = ESat.UNDEFINED;
+        if (isCompletelyInstantiated()) {
+            double lb = yVar.getLB() - Math.pow(xVar.getLB(), k);
+            double ub = yVar.getUB() - Math.pow(xVar.getUB(), k);
+            lb = RealUtils.roundValue(lb, xVar.getPrecision());
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:56]: Why not calling `RealUtils.prevFloat` ?

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            ub = RealUtils.roundValue(ub, xVar.getPrecision());
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:57]: Why not calling `RealUtils.nextFloat` ?

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        boolean hasChanged = true;
+        // Projection + propagation
+        while (hasChanged) {
+            hasChanged = kPower();
+            hasChanged |= kRoot();
+        }
+    }
+
+    private boolean kPower() throws ContradictionException {
+        RealInterval x = new RealIntervalConstant(xVar.getLB(), xVar.getUB());
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:74]: `xVar` extends `RealInterval`, creating a `RealIntervalConstant ` is not necessary.

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        RealInterval y = new RealIntervalConstant(yVar.getLB(), yVar.getUB());
+        RealInterval power = RealUtils.iPower(x, k); // y = x^k
+        // For even power, y will range [0, +Inf]
+        // For odd power, y will range [-Inf, +Inf]
+        // So it doesn't require special handling to update the bounds
+        y = RealUtils.updateBounds(y, power);
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:80]: Previous lines, including this one, can be replaced by:
```java
RealInterval y = RealUtils.updateBounds(yVar, RealUtils.iPower(xVar, k));
```

File: solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        return yVar.updateBounds(y.getLB(), y.getUB(), this);
+    }
+
+    private boolean kRoot() throws ContradictionException {
+        boolean isIntervalPositive = xVar.getLB() >= 0;
+        boolean isIntervalNegative = xVar.getUB() <= 0;
+        RealInterval x = new RealIntervalConstant(xVar.getLB(), xVar.getUB());
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/constraints/real/RealPowerPropagator.java:87]: The method can also be simplified, using already existant API:
```java
boolean isIntervalPositive = xVar.getLB() >= 0;
boolean isIntervalNegative = xVar.getUB() <= 0;
RealInterval root = RealUtils.iRoot(yVar, k); // x = root(y, k)
boolean isEven = k % 2 == 0;
// In all cases, bounds ar calculate as intersection between x and temp
if (isEven) {
    if (isIntervalPositive) {
        return xVar.updateBounds(root.getLB(), root.getUB(), this);
    } else if (isIntervalNegative) {
        return xVar.updateBounds(-root.getUB(), -root.getLB(), this);
    } else { // It's crossing zero
        return xVar.updateBounds(-root.getUB(), root.getUB(), this);
    }
} else {
    // For odd roots isn't necessary specific handling as roots can be negative naturally
    return xVar.updateBounds(root.getLB(), root.getUB(), this);
}
```

File: solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/CstCArExpression.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
+package org.chocosolver.solver.expression.continuous.arithmetic;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/CstCArExpression.java:1]: Why this new classe is needed?
I think everything can be moved to `BiCArExpression`.

File: solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/CstCArExpression.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        if (me == null) {
+            RealVar xVar = e.realVar(p);
+            switch (op) {
+                case POW:
+                    if (isInteger(constant)) {
+                        int k = (int) constant;
+                        if (constant >= 0) {
+                            // y = x^k
+                            RealInterval powerResult = RealUtils.iPower(xVar, k);
+                            me = model.realVar(powerResult.getLB(), powerResult.getUB(), xVar.getPrecision());
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/CstCArExpression.java:50]: I'm quite uncomfortable with this. 
The `RealVar realVar(double p)` is supposed to deal with stuff for Ibex. But here, you are doing something different. 
If we want this to be coherent with wath already exists, I would suggest to move the constraint declaration in `IRealConstraintFactory`. 
The power constraint will then be called directly from there.

File: solver/src/main/java/org/chocosolver/util/tools/RealUtils.java [hunk 13]
```diff
@@ -729,4 +731,24 @@ public static RealInterval acos_wrt(RealInterval i, RealInterval res) {
         return new RealIntervalConstant(retInf, retSup);
     }
 
+    public static RealInterval updateBounds(RealInterval var, RealInterval newInterval) {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/util/tools/RealUtils.java:734]: this method already exists for `RealVar` and it not required anymore with the correction I suggested.

File: solver/src/main/java/org/chocosolver/util/tools/RealUtils.java [hunk 14]
```diff
@@ -729,4 +731,24 @@ public static RealInterval acos_wrt(RealInterval i, RealInterval res) {
...
+        return updateBounds(var, newInterval.getLB(), newInterval.getUB());
+    }
+
+    public static RealInterval updateBounds(RealInterval var, double lb, double ub) {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/util/tools/RealUtils.java:738]: Same comment as previous one.

File: solver/src/main/java/org/chocosolver/util/tools/RealUtils.java [hunk 15]
```diff
@@ -729,4 +731,24 @@ public static RealInterval acos_wrt(RealInterval i, RealInterval res) {
...
+        return new RealIntervalConstant(
+                Math.max(var.getLB(), lb),
+                Math.min(var.getUB(), ub));
+    }
+
+    public static double roundValue(double value, double precision) {
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/util/tools/RealUtils.java:744]: That seems strange to me to pass by a `String` to get a double.
I think calling `prevFloat` and `nextFloat` are enough in `isEntailed`

File: solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/CArExpression.java [hunk 16]
```diff
@@ -316,7 +316,7 @@ default CArExpression div(CArExpression y) {
      * @return return the expression "x^y" where this is "x"
      */
     default CArExpression pow(double y) {
-        return new BiCArExpression(CArExpression.Operator.POW, this, this.getModel().realVar(y));
+        return new CstCArExpression(CArExpression.Operator.POW, this, y);
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/CArExpression.java:319]: As I said, I don't think `CstCArExpression` is needed.

File: solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/UnCArExpression.java [hunk 17]
```diff
@@ -48,7 +48,7 @@
     /**
      * The expression this expression relies on
      */
-    private CArExpression e;
+    CArExpression e;
```
Comment (cprudhom) [inline solver/src/main/java/org/chocosolver/solver/expression/continuous/arithmetic/UnCArExpression.java:51]: I believe this modification is not needed anymore if `CstCArExpression` disappears.

**Review by cprudhom — CHANGES_REQUESTED** [review]
Thank you for contributing.
Here are my general comments.
1. Nice test suite, thanks!
2. The propagator is OK to me, but I do not agree with the way it is used:
- `RealVar realVar(double p)` is supposed to be used exclusively with Ibex
- Commonly, mixed constraints (ie, continuous and discrete) are declared in `IRealConstraintFactory`.
Consequently, I think the code should be revamped.
3. I don't think adding new methods to `RealUtils` is required too.

**Comment by schmittjoaopedro** [comment]
All right, I see all your points and I will revise all of them.
However right now I'm very busy with other stuffs, but in the meantime I was thinking in create another pull request to just contribute with the test suite. I think it's very important and don't want to let it in standby (outside from the master branch) until I came back to continue the development.

What do you think?

**Comment by cprudhom** [comment]
Sure, you can split the PR in 2

**Comment by schmittjoaopedro** [comment]
I've created the PR #723 to segregate the test suite and propose some improvements for ibex integration.

About this PR, I was wondering and I agree that it's not a good idea to replace the ibex integration with a new propagator, because it would reduce the expected homogeneous behavior of IBEX. I think that a better solution should be follow the guidelines of the H4C algorithm implementing the `tighten` and `project` methods.

I'll not have time to think about that now, so I will close this PR for while.

**Pull request closed without merging**